### PR TITLE
Adds e2e Upgrade Test

### PR DESCRIPTION
**Pull request by danehans**
Adds an e2e upgrade test that performs the following:
- Sets the operator to the latest release to simulate the upgrade "from" release.
- Provisions a `Contour` using release "latest" and verifies e2e connectivity, no errors in operator logs, etc..
- Upgrades the operator to simulate the upgrade "to" release.
- Ensures the operator and Contour are using the updated image tag.
- Verifies post-upgrade e2e connectivity.

Partially fixes https://github.com/projectcontour/contour-operator/issues/142. Due to a) PR size and b) current Gateway API implementation under review, this PR does not include Gateway upgrade testing.

Signed-off-by: Daneyon Hansen <daneyonhansen@gmail.com>

### CODE REVIEW

**Comment by danehans** [comment]
[test results](https://gist.githubusercontent.com/danehans/7d2cfe0f74b59e71fd92275bd9d0d811/raw/1f0e0492e3a5e64c9bbb158c896cada43756fb9e/contour_operator_upgrade_test.md)

File: test/e2e/operator_test.go [hunk 1]
```diff
@@ -876,3 +876,173 @@ func TestGatewayOwnership(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("failed to create contour %s/%s: %v", operatorNs, contourName, err)
+	}
+	t.Logf("created contour %s/%s", cntr.Namespace, cntr.Name)
+
+	// The contour should now report available.
+	if err := waitForContourStatusConditions(ctx, kclient, 3*time.Minute, contourName, operatorNs, expectedContourConditions...); err != nil {
+		t.Fatalf("failed to observe expected status conditions for contour %s/%s: %v", operatorNs, testName, err)
+	}
+	t.Logf("observed expected status conditions for contour %s/%s", testName, operatorNs)
```
Comment (skriss) [inline test/e2e/operator_test.go:927]:
```suggestion
	t.Logf("observed expected status conditions for contour %s/%s", operatorNs, contourName)
```

File: test/e2e/operator_test.go [hunk 2]
```diff
@@ -876,3 +876,173 @@ func TestGatewayOwnership(t *testing.T) {
...
+
+	cntr, err := newContour(ctx, kclient, cfg)
+	if err != nil {
+		t.Fatalf("failed to create contour %s/%s: %v", operatorNs, contourName, err)
+	}
+	t.Logf("created contour %s/%s", cntr.Namespace, cntr.Name)
+
+	// The contour should now report available.
+	if err := waitForContourStatusConditions(ctx, kclient, 3*time.Minute, contourName, operatorNs, expectedContourConditions...); err != nil {
+		t.Fatalf("failed to observe expected status conditions for contour %s/%s: %v", operatorNs, testName, err)
```
Comment (skriss) [inline test/e2e/operator_test.go:925]:
```suggestion
		t.Fatalf("failed to observe expected status conditions for contour %s/%s: %v", operatorNs, contourName, err)
```

**Review by skriss — CHANGES_REQUESTED** [review]

**Comment by danehans** [comment]
@skriss thanks for the review. Commit `d4e047d` resolves your comments above.

**Review by skriss — APPROVED** [review]
LGTM

File: test/e2e/operator_test.go [hunk 3]
```diff
@@ -876,3 +876,173 @@ func TestGatewayOwnership(t *testing.T) {
...
+
+	// Get the Envoy ClusterIP to curl.
+	svcName := "envoy"
+	ip, err := envoyClusterIP(ctx, kclient, specNs, svcName)
+	if err != nil {
+		t.Fatalf("failed to get clusterIP for service %s/%s: %v", specNs, svcName, err)
+	}
+
+	// Curl the ingress from the client pod.
+	url := fmt.Sprintf("http://%s/", ip)
```
Comment (sunjayBhatia) [inline test/e2e/operator_test.go:971]: How come this was done with curl in a pod rather than just from the test/go like other tests?

Comment (danehans) [reply test/e2e/operator_test.go:971]: @sunjayBhatia b/c the Contour is [configured](https://github.com/projectcontour/contour-operator/pull/316/files#diff-cf4b4e5424070a666a364d1d1b04011478d888d6d3d65c943ca7783333b7a6e4R914) with a "ClusterIPService" network publishing type. Therefore the Envoy Service is only reachable from within the cluster.

Comment (sunjayBhatia) [reply test/e2e/operator_test.go:971]: ah gotcha, missed that 👍🏽

**Review by sunjayBhatia — APPROVED** [review]

**Pull request merged**